## 🛠️ **ETL (Extract, Transform, Load)**



#### **📂 Procesamiento del 1er archivo: `Yelp/business.pkl`**

####  **Importamos las librerías que vamos a usar**


In [1]:
import pandas as pd
import data_utils   
from data_utils import data_type_check, data_type_check_pkl

import warnings
warnings.filterwarnings('ignore')
import seaborn as sns
import matplotlib.pyplot as plt


#### 📦 **Extraccion** de los datos y primera exploración 


Mostrar info como en datatypecheck
#data_type_check_pkl('../0_Dataset/Yelp/business.pkl')

**PENDIENTE**
% de nulos, .shape, .columns


In [2]:
business = pd.read_pickle('../0_Dataset/Yelp/business.pkl')
business.info()
business.sample(2)

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 28 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
 14  business_id   5 non-null       object
 15  name          5 non-null       object
 16  address       5 non-null       object
 17  city          5 non-null       object
 18  state         5 non-null     

business_id                                name  \
96751  avZoFq_TQZQIjR8PX7PgTg                              Pier 1   
9201   vyXNBZY_pG4Fk5TtFm9o8Q  Burger & Company - Old Hickory, TN   

                    address         city state postal_code   latitude  \
96751  128 S Westshore Blvd        Tampa    TN       33609  27.943913   
9201         700 Hadley Ave  Old Hickory    LA       37138  36.263946   

       longitude stars review_count  ... state postal_code latitude longitude  \
96751 -82.523481   3.5           10  ...   NaN         NaN      NaN       NaN   
9201  -86.648122   4.0           42  ...   NaN         NaN      NaN       NaN   

      stars review_count is_open attributes categories hours  
96751   NaN          NaN     NaN        NaN        NaN   NaN  
9201    NaN          NaN     NaN        NaN        NaN   NaN  

[2 rows x 28 columns]

Tenemos columnas duplicadas, procedemos a quitarlas.


In [3]:
business = business.loc[:, ~business.columns.duplicated()]
business.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150346 entries, 0 to 150345
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   business_id   150346 non-null  object
 1   name          150346 non-null  object
 2   address       150346 non-null  object
 3   city          150346 non-null  object
 4   state         150343 non-null  object
 5   postal_code   150346 non-null  object
 6   latitude      150346 non-null  object
 7   longitude     150346 non-null  object
 8   stars         150346 non-null  object
 9   review_count  150346 non-null  object
 10  is_open       150346 non-null  object
 11  attributes    136602 non-null  object
 12  categories    150243 non-null  object
 13  hours         127123 non-null  object
dtypes: object(14)
memory usage: 17.2+ MB


#### 🔁 **TRANSFORM**

Nos interesa observar las categorías por lo que primero vamos a corroborar que no tenga duplicados

In [4]:
business.categories.value_counts()

categories
Beauty & Spas, Nail Salons                                                                                       1012
Restaurants, Pizza                                                                                                935
Nail Salons, Beauty & Spas                                                                                        934
Pizza, Restaurants                                                                                                823
Restaurants, Mexican                                                                                              728
                                                                                                                 ... 
Dermatologists, Health & Medical, Cosmetic Surgeons, Doctors, Acne Treatment, Skin Care, Beauty & Spas              1
Home Services, Home & Garden, Nurseries & Gardening, Hardware Stores, Shopping, Building Supplies, Appliances       1
Food Trucks, Smokehouse, Restaurants, Food, B

Verificamos los nulos y duplicados de categories

In [5]:
# Cuenta los duplicados y nulos en la columna 'categories' antes de cualquier eliminación
duplicados_inicial = business['categories'].duplicated().sum()
nulos_inicial = business['categories'].isna().sum()

print("Antes: Cantidad de duplicados en 'categories':", duplicados_inicial)
print("Antes: Cantidad de NaNs en 'categories':", nulos_inicial)

# Elimina las filas con valores NaN en la columna 'categories'
business = business.dropna(subset=['categories'])

# Cuenta los duplicados y nulos en la columna 'categories' después de eliminar NaNs
duplicados_despues_nulos = business['categories'].duplicated().sum()
nulos_despues_nulos = business['categories'].isna().sum()

print("Después: Cantidad de NaNs en 'categories' ", nulos_despues_nulos)


'''
________________________________________________________________
FUTURO: Descomentar o borrar
Elimina las filas duplicadas basadas en la columna 'categories'
business = business.drop_duplicates(subset=['categories'])  
_________________________________________________________________  
'''

# Cuenta los duplicados en la columna 'categories' después de eliminar duplicados
duplicados_final = business['categories'].duplicated().sum()

print("Después: Cantidad de duplicados en 'categories'", duplicados_final)


Antes: Cantidad de duplicados en 'categories': 67185
Antes: Cantidad de NaNs en 'categories': 103
Después: Cantidad de NaNs en 'categories'  0
Después: Cantidad de duplicados en 'categories' 67083


Dado que diferentes negocios pueden compartir las mismas 'categories', no borramos los  67083 duplicados 


### Filtramos por estado y categoria


Vamos a filtrar por estado. Para eso primero observamos todos los estados y luego elejimos el que vamos a usar

In [6]:
business['state'].value_counts()

state
PA     34013
FL     26311
TN     12045
IN     11238
MO     10909
LA      9917
AZ      9907
NJ      8529
NV      7712
AB      5571
CA      5201
ID      4463
DE      2261
IL      2142
TX         4
CO         3
WA         2
HI         2
MA         2
NC         1
UT         1
MT         1
MI         1
SD         1
XMS        1
VI         1
VT         1
Name: count, dtype: int64

Vamos a observar los negocios de Pennsylvania

In [7]:
# Se observan los negocios de California
df_estado_PA = business[business['state']=='PA']

df_estado_PA['categories'].value_counts()

categories
Nail Salons, Beauty & Spas                                                                                             227
Beauty & Spas, Nail Salons                                                                                             218
Restaurants, Pizza                                                                                                     206
Pizza, Restaurants                                                                                                     177
Restaurants, Chinese                                                                                                   170
                                                                                                                      ... 
Cafes, Breakfast & Brunch, Event Planning & Services, Caterers, Restaurants, Salad                                       1
Casinos, Hostels, Arts & Entertainment, Comedy Clubs, Hotels & Travel, Nightlife, Event Planning & Services, Hotels      1
Local

Observamos que en Pennsylvania el top 3 categorias es:
- Nail Salons, Beauty & Spas : 227
- Beauty & Spas, Nail Salons: 218
- Restaurants, Pizza: 206
- Pizza, Restaurants: 177

In [8]:
'''
Unir categorias?
'''

'\nUnir categorias?\n'

Filtramos por restaurantes en Pennsylvania

In [9]:
# Cambia el tipo de dato de categories a string
business['categories'] = business['categories'].astype(str)

# Filtra las filas que contienen 'restaurant' en 'categories' en Pennsylvania
restaurante_pa = business[business['categories'].str.lower().str.contains('restaurant', na=False) & (business['state'] == 'PA')].reset_index(drop=True)

# Cuenta las ocurrencias de las categorías
conteo_categorias_pa = restaurante_pa['categories'].value_counts()
print(conteo_categorias_pa)


categories
Restaurants, Pizza                                                                   206
Pizza, Restaurants                                                                   177
Restaurants, Chinese                                                                 170
Mexican, Restaurants                                                                 151
Restaurants, Mexican                                                                 149
                                                                                    ... 
Restaurants, Pizza, Pubs, Bars, Nightlife, Italian                                     1
Latin American, Restaurants, Mexican, Dominican, Caribbean                             1
American (Traditional), Burgers, Restaurants, Tapas/Small Plates                       1
Restaurants, Cheesesteaks, Pizza                                                       1
Restaurants, Comfort Food, Food, Food Trucks, Caterers, Event Planning & Services      1
Name: coun

In [10]:
restaurante_pa.columns

Index(['business_id', 'name', 'address', 'city', 'state', 'postal_code',
       'latitude', 'longitude', 'stars', 'review_count', 'is_open',
       'attributes', 'categories', 'hours'],
      dtype='object')

Filtramos los restaurantes a traves de su latitud y longitud, seleccionando solamente los que se encuentren abiertos y en Pennsylvania

In [11]:
# Define los límites geográficos para Pennsylvania
max_latitud = 42.3
min_latitud = 39.7
max_longitud = -74.4
min_longitud = -75.3

# Se establecen las máscaras para filtrar
msk_lat = (restaurante_pa['latitude'].astype(float) >= min_latitud) & (restaurante_pa['latitude'].astype(float) <= max_latitud)
msk_lon = (restaurante_pa['longitude'].astype(float) >= min_longitud) & (restaurante_pa['longitude'].astype(float) <= max_longitud)

# Filtra el DataFrame `restaurante_pa` por coordenadas geográficas
restaurante_pa = restaurante_pa[msk_lat & msk_lon].reset_index(drop=True)

restaurante_pa= restaurante_pa[restaurante_pa['is_open']==1].reset_index(drop=True)
restaurante_pa.drop(columns=['is_open','state'], inplace=True)

# Verifica el DataFrame resultante
restaurante_pa.head(2)

business_id              name       address          city state  \
0  QdN72BWoyFypdGJhhI5r7g           Bar One  767 S 9th St  Philadelphia    PA   
1  Mjboz24M9NlBeiOJKLEd_Q  DeSandro on Main  4105 Main St  Philadelphia    PA   

  postal_code   latitude  longitude stars review_count is_open  \
0       19147  39.939825 -75.157447   4.0           65       0   
1       19127  40.022466 -75.218314   3.0           41       0   

                                          attributes  \
0  {'Smoking': 'u'no'', 'NoiseLevel': 'u'average'...   
1  {'RestaurantsReservations': 'False', 'Caters':...   

                                          categories  \
0  Cocktail Bars, Bars, Italian, Nightlife, Resta...   
1                    Pizza, Restaurants, Salad, Soup   

                                               hours  
0  {'Monday': '16:0-0:0', 'Tuesday': '16:0-0:0', ...  
1  {'Tuesday': '17:0-21:30', 'Wednesday': '17:0-1...

In [12]:
# Verifica el DataFrame resultante
data_type_check(restaurante_pa)
restaurante_pa.head(2)


 Resumen del dataframe:

Dimensiones:  (2698, 14)
         columna  %_no_nulos  %_nulos  total_nulos tipo_dato
0    business_id      100.00     0.00            0    object
1           name      100.00     0.00            0    object
2        address      100.00     0.00            0    object
3           city      100.00     0.00            0    object
4          state      100.00     0.00            0    object
5    postal_code      100.00     0.00            0    object
6       latitude      100.00     0.00            0    object
7      longitude      100.00     0.00            0    object
8          stars      100.00     0.00            0    object
9   review_count      100.00     0.00            0    object
10       is_open      100.00     0.00            0    object
11    attributes       98.93     1.07           29    object
12    categories      100.00     0.00            0    object
13         hours       83.47    16.53          446    object


business_id              name       address          city state  \
0  QdN72BWoyFypdGJhhI5r7g           Bar One  767 S 9th St  Philadelphia    PA   
1  Mjboz24M9NlBeiOJKLEd_Q  DeSandro on Main  4105 Main St  Philadelphia    PA   

  postal_code   latitude  longitude stars review_count is_open  \
0       19147  39.939825 -75.157447   4.0           65       0   
1       19127  40.022466 -75.218314   3.0           41       0   

                                          attributes  \
0  {'Smoking': 'u'no'', 'NoiseLevel': 'u'average'...   
1  {'RestaurantsReservations': 'False', 'Caters':...   

                                          categories  \
0  Cocktail Bars, Bars, Italian, Nightlife, Resta...   
1                    Pizza, Restaurants, Salad, Soup   

                                               hours  
0  {'Monday': '16:0-0:0', 'Tuesday': '16:0-0:0', ...  
1  {'Tuesday': '17:0-21:30', 'Wednesday': '17:0-1...


 Resumen del dataframe:

Dimensiones:  (2698, 14)
         columna  %_no_nulos  %_nulos  total_nulos tipo_dato
0    business_id      100.00     0.00            0    object
1           name      100.00     0.00            0    object
2        address      100.00     0.00            0    object
3           city      100.00     0.00            0    object
4          state      100.00     0.00            0    object
5    postal_code      100.00     0.00            0    object
6       latitude      100.00     0.00            0    object
7      longitude      100.00     0.00            0    object
8          stars      100.00     0.00            0    object
9   review_count      100.00     0.00            0    object
10       is_open      100.00     0.00            0    object
11    attributes       98.93     1.07           29    object
12    categories      100.00     0.00            0    object
13         hours       83.47    16.53          446    object


Vamos a filtrar los restaurantes abiertos de los cerrados


# EN PROCESO

In [ ]:
#restaurante_estado= restaurante_estado[restaurante_estado['is_open']==1].reset_index(drop=True)
#data_type_check(restaurante_estado)

In [ ]:
#restaurante_estado.categories.value_counts()

#### **📤 LOAD**

In [ ]:
#guardar en parquet